servicetalk-http-utils/src/main/java/io/servicetalk/http/utils/TimeoutHttpRequesterFilter.java

```diff
@@ -30,46 +32,161 @@
+        this.timeoutExecutor = null;
+    }
+
+    /**
+     * Creates a new instance.
+     *
+     * @param timeoutForRequest function for extracting timeout from request which may also determine the timeout using
+     * other sources. If no timeout is to be applied then the function should return null.
+     * @param fullRequestResponse if {@code true} then timeout is for full request/response transaction otherwise only
+     * the response metadata must arrive before the timeout.
+     * @param timeoutExecutor the {@link Executor} to use for managing the timer notifications
+     */
+    public TimeoutHttpRequesterFilter(final TimeoutFromRequest timeoutForRequest,
+                                      final boolean fullRequestResponse,
+                                      final Executor timeoutExecutor) {
+        this.timeoutForRequest = Objects.requireNonNull(timeoutForRequest, "timeoutForRequest");
+        this.fullRequestResponse = fullRequestResponse;
+        this.timeoutExecutor = Objects.requireNonNull(timeoutExecutor, "timeoutExecutor");
+    }
+
+    /**
+     * Returns a function which returns the provided default duration as the timeout duration to be used for any
+     * request.
+     *
+     * @param duration timeout duration or null for no timeout
+     * @return a function to produce a timeout using specified duration
+     */
+    static TimeoutFromRequest simpleDurationTimeout(@Nullable Duration duration) {
+        return new TimeoutFromRequest() {
+            @Nullable
+            @Override
+            public Duration apply(final HttpRequestMetaData request) {
+                // the request is not considered
+                return duration;
+            }
+
+            @Override
+            public HttpExecutionStrategy influenceStrategy(final HttpExecutionStrategy strategy) {
+                // No influence since we do not block.
+                return strategy;
+            }
+        };
     }
 
     private Single<StreamingHttpResponse> request(final StreamingHttpRequester delegate,
                                                   final HttpExecutionStrategy strategy,
                                                   final StreamingHttpRequest request) {
-        return timeoutExecutor != null ? delegate.request(strategy, request).timeout(duration, timeoutExecutor) :
-                delegate.request(strategy, request).timeout(duration);
+        return Single.defer(() -> {
+            Single<StreamingHttpResponse> response = delegate.request(strategy, request);
+
+            Duration timeout = timeoutForRequest.apply(request);
+
+            if (null != timeout) {
+                Single<StreamingHttpResponse> timeoutResponse = timeoutExecutor != null ?
+                        response.timeout(timeout, timeoutExecutor) : response.timeout(timeout);
+
+                response = fullRequestResponse ?
+                        Single.defer(() -> {
```

Comment:
You are already inside the `Single.defer` that handles `Single<StreamingHttpResponse>`, there is not need for another one

Reply:
thank you.